tests/tools/oesign-engine/test-engine/oesign-test-engine.c

```diff
@@ -0,0 +1,275 @@
+     printf("test_engine_load_pubkey engine=%p, keyid = %s\n",  e, keyid);
+     return test_engine_only_pubkey;
+}
+
+
+static int test_engine_init(ENGINE *e)
+{ 
+    int ret = 0;
+
+    printf("test_engine_init\n");
+
+    test_engine_only_privkey = EVP_PKEY_new();
+    test_engine_only_pubkey  = EVP_PKEY_new();
+    if (!ENGINE_set_load_privkey_function(e, test_engine_load_privkey)){
+        fprintf(stderr, "ENGINE_set_id failed\n");
+        goto done;
+    }
+    if (!ENGINE_set_load_pubkey_function(e, test_engine_load_pubkey)) {
+        fprintf(stderr, "ENGINE_set_id failed\n");
+        goto done;
+    }
+
+    ret = 1;
+
+done:
+    return ret;
+}
+
+static int test_engine_finish(ENGINE *e)
+{ 
+    (void)e;
+    printf("test_engine_finish\n");
+    return 1;
+}
+
+static int test_engine_destroy(ENGINE *e)
+{ 
+    (void)e;
+    printf("test_engine_destroy\n");
+    //2do:test_engine_param_free();
+    EVP_PKEY_free(test_engine_only_privkey);
+    EVP_PKEY_free(test_engine_only_pubkey);
+    return 1;
+}
+
+#if 0
+static int test_engine_pkey_meths(ENGINE *e, EVP_PKEY_METHOD **pmeth,
+                           const int **pnids, int nid)
+{
+    static const int rnid = EVP_PKEY_RSA;
+    if (pmeth == NULL) {
+        *pnids = &rnid;
+        return 1;
+    }
+
+    if (nid == EVP_PKEY_RSA) {
+        *pmeth = test_engine_rsa;
+        return 1;
+    }
+
+    *pmeth = NULL;
+    return 0;
+}
+#endif
+
+
+static int test_engine_bind(ENGINE *e, const char *id)
+{
+    int ret = 0;
+
+    if (!ENGINE_set_id(e, test_engine_id)) {
+        fprintf(stderr, "ENGINE_set_id failed\n");
+        goto done;
+    }
+    if (!ENGINE_set_name(e, test_engine_name)) {
+        printf("ENGINE_set_name failed\n");
+        goto done;
+    }
+
+#if 0
+    if (!ENGINE_set_digests(e, test_engine_digests)) {
+        printf("ENGINE_set_digests failed\n");
+        goto done;
+    }
+    if (!ENGINE_set_ciphers(e, test_engine_ciphers)) {
+        printf("ENGINE_set_ciphers failed\n");
+        goto done;
+    }
+    if (!ENGINE_set_pkey_meths(e, test_engine_pkey_meths)) {
+        printf("ENGINE_set_pkey_meths failed\n");
+        goto done;
+    }
+    if (!ENGINE_set_pkey_asn1_meths(e, test_engine_pkey_asn1_meths)) {
+        printf("ENGINE_set_pkey_asn1_meths failed\n");
+        goto done;
+    }
+
+    /* Control function and commands */
+    if (!ENGINE_set_cmd_defns(e, test_engine_cmds)) {
+        fprintf(stderr, "ENGINE_set_cmd_defns failed\n");
+        goto done;
+    }
+    if (!ENGINE_set_ctrl_function(e, test_engine_control_func)) {
+        fprintf(stderr, "ENGINE_set_ctrl_func failed\n");
+        goto done;
+    }
+#endif
+
+    if (!ENGINE_set_destroy_function(e, test_engine_destroy)
+        || !ENGINE_set_init_function(e, test_engine_init)
+        || !ENGINE_set_finish_function(e, test_engine_finish)) {
+        goto done;
+    }
+
+
+    ret = 1;
+
+    printf("zoot id=%p\n", id);
+    printf("zoot id=%s\n", id);
```

Comment:
output looks odd.  Two visually different values with same label?  If this is still needed, improve the labels.

Reply:
debug printfs got past. Sorry